homeassistant/components/esphome/sensor.py

```diff
@@ -71,9 +74,84 @@ async def async_setup_entry(
 )
 
 
-class EsphomeSensor(EsphomeEntity[SensorInfo, SensorState], SensorEntity):
+class EsphomeSensor(
+    EsphomeEntity[SensorInfo, SensorState], SensorEntity, RestoreEntity
+):
     """A sensor implementation for esphome."""
 
+    async def async_added_to_hass(self) -> None:
+        """Register callbacks."""
+        await super().async_added_to_hass()
+
+        if self._static_info.last_reset_type != LastResetType.AUTO:
+            return
+
+        # Logic to restore old state for last_reset_type AUTO:
+        last_state = await self.async_get_last_state()
+        if last_state is None:
+            return
+
+        if "last_reset" in last_state.attributes:
+            self._attr_last_reset = datetime.fromisoformat(
+                last_state.attributes["last_reset"]
+            )
+
+        try:
+            self._old_state = float(last_state.state)
+        except ValueError:
+            pass
+
+    _old_state: float | None = None
+
+    @callback
+    def _on_state_update(self) -> None:
+        """Check last_reset when new state arrives."""
+        if self._static_info.last_reset_type == LastResetType.NEVER:
+            self._attr_last_reset = datetime.min
+
+        if self._static_info.last_reset_type != LastResetType.AUTO:
+            super()._on_state_update()
+            return
+
+        # Last reset type AUTO logic for the last_reset property
+        # In this mode we automatically determine if an accumulator reset
+        # has taken place.
+        # We compare the last valid value (_old_state) with the new one.
+        # If the value has reset to 0 or has significantly reduced we say
+        # it has reset.
+        new_state: float | None = None
+        state = cast("str | None", self.state)
+        if state is not None:
+            try:
```

Comment:
We can use `contextlib.suppress`.
https://docs.python.org/3/library/contextlib.html#contextlib.suppress